src/test/java/net/openhft/chronicle/core/threads/ThreadLockTest.java

```diff
@@ -0,0 +1,267 @@
+        }
+        value.getValues.add(deadThreadId << 32);
+        value.getValues.add(deadThreadId << 32);
+
+        lock.lock(1);
+        assertEquals(0, value.getValues.size());
+
+        assertEquals("8000000000000001", value.setValues.stream().map(Long::toHexString).collect(Collectors.joining(", ")));
+    }
+
+    @Test
+    public void interrupted() {
+        final VanillaLongValue value = new VanillaLongValue();
+        ThreadLock lock = new ThreadLock(value, 100);
+        final Thread thread = Thread.currentThread();
+        thread.interrupt();
+        lock.lock(2);
+        try {
+            lock.lock(1);
+            fail();
+        } catch (InterruptedRuntimeException ire) {
+            // expected
+        }
+        assertTrue(thread.isInterrupted());
+    }
+
+    @Test(expected = IllegalArgumentException.class)
+    public void invalidThreadId() {
+        final VanillaLongValue value = new VanillaLongValue();
+        ThreadLock lock = new ThreadLock(value, 100);
+        lock.lock(0);
+    }
+
+    static class VanillaLongValue implements LongValue {
+        long value;
+
+        @Override
+        public long getValue() throws IllegalStateException {
+            return value;
+        }
+
+        @Override
+        public void setValue(long value) throws IllegalStateException {
+            this.value = value;
+        }
+
+        @Override
+        public long getVolatileValue() throws IllegalStateException {
+            return getValue();
+        }
+
+        @Override
+        public void setVolatileValue(long value) throws IllegalStateException {
+            setValue(value);
+        }
+
+        @Override
+        public void setOrderedValue(long value) throws IllegalStateException {
+            setValue(value);
+        }
+
+        @Override
+        public long addValue(long delta) throws IllegalStateException {
+            throw new AssertionError();
+        }
+
+        @Override
+        public long addAtomicValue(long delta) throws IllegalStateException {
+            throw new AssertionError();
+        }
+
+        @Override
+        public boolean compareAndSwapValue(long expected, long value) throws IllegalStateException {
+            if (getValue() == expected) {
+                setValue(value);
+                return true;
+            }
+            return false;
+        }
+
+        @Override
+        public String toString() {
+            return getClass().getName() + "@" + Integer.toHexString(hashCode());
+        }
+    }
+
+    static class CheckingLongValue extends VanillaLongValue {
+        List<Long> getValues = new ArrayList<>();
+        List<Long> setValues = new ArrayList<>();
+
+        @Override
+        public long getValue() throws IllegalStateException {
+            return getValues.remove(0);
+        }
+
+        @Override
+        public void setValue(long value) throws IllegalStateException {
+            setValues.add(value);
+        }
+    }
```

Comment:
Would be good to see a stress test that spun up a bunch of threads that contended for a lock and checked some guarded state to ensure multiple threads didn't get in. Including some that died holding the lock.

Reply:
And a multi-process version might be important as well

Reply:
A test well worth adding. I did one performance test and it showed an issue.